### PR TITLE
Add basic auth to curl download strategy, HOMEBREW_DOCKER_REGISTRY_TO…

### LINKED ISSUE #12574: Add basic auth capability for private Docker registries
### Provide a detailed description of the proposed feature

Today, thanks to https://github.com/Homebrew/brew/pull/11766, only bearer tokens are supported to authenticate against a private Docker registry. Configuration example:

```bash
export HOMEBREW_ARTIFACT_DOMAIN=https://artifactory.your-company.com/artifactory/brew-remote
export HOMEBREW_DOCKER_REGISTRY_TOKEN=<artifactory-access-token>
````

I propose to implement the support of basic authentication. Configuration example of my proposal:

```bash
export HOMEBREW_ARTIFACT_DOMAIN=https://artifactory.your-company.com/artifactory/brew-remote
export HOMEBREW_DOCKER_REGISTRY_BASIC_USER=<username>
export HOMEBREW_DOCKER_REGISTRY_BASIC_PWD=<password>
````

... which would send an `Authorization: Basic ...`  header instead of a Bearer one:
https://github.com/Homebrew/brew/blob/371122da557d770aa4555f258a904aa008f9ce8b/Library/Homebrew/download_strategy.rb#L599

### What is the motivation for the feature?

In our case we use Artifactory as a mirror. It supports both authentication methods:

- [access token](https://www.jfrog.com/confluence/display/JFROG/Access+Tokens) (Bearer)
- [username / API key](https://www.jfrog.com/confluence/display/JFROG/User+Profile#UserProfile-APIKey) (Basic)

However, only super admins can generate access tokens that don't expire.
Regular users can only generate access tokens with a 3600s duration maximum.
On the other hand, regular users can generate a random API key that can be used as a password in a basic auth that doesn't expire.

Also, many other Docker registries only support basic auth.

### How will the feature be relevant to at least 90% of Homebrew users?

The support of this feature will really help using homebrew in large companies where you are behind a proxy and you are forced to use a mirror like Artifactory with restricted permissions.

### What alternatives to the feature have been considered?

Today we developed this wrapper around homebrew that automatically refreshes Artifactory's access token when it's expired.
The dev experience becomes quite decent, but it's still hacky.

Snippet to include in your `~/.zshrc` or equivalent:

```bash
export HOMEBREW_ARTIFACT_DOMAIN=https://artifactory.your-company.com/artifactory/brew-remote

brewtoken() {
  if [[ -f ~/.brewtoken.sh ]]; then
    source ~/.brewtoken.sh
  fi

  if [[ ! -z "${BREWTOKEN_TS:-}" && "${1:-}" != "--force" ]]; then
    if [[ $(( $(date +%s) - $BREWTOKEN_TS )) -le 3300 ]]; then
      # Token not expired > we silently exit
      return
    fi
  fi

  >&2 echo "Requesting access token from Artifactory..."
  local rtRes=$(jfrog rt access-token-create)
  local rtAccessToken=$(echo "${rtRes}" | tr -d ' ' | grep -o '"access_token":"[^"]*"' | tr -d \" | sed 's/^access_token://') # we cannot use jq because we need homebrew to install it :-)
  if [[ -z "${rtAccessToken}" ]]; then
    >&2 echo "Failed to fetch access token:"
    >&2 echo "${rtRes}"
    return 1
  fi

  echo "export BREWTOKEN_TS=$(date +%s)" > ~/.brewtoken.sh
  echo "export HOMEBREW_DOCKER_REGISTRY_TOKEN=\"${rtAccessToken}\"" >> ~/.brewtoken.sh
  source ~/.brewtoken.sh
}

brew() {
  if [[ "${1:-}" =~ 'install|upgrade|reinstall|fetch|formulae|search|update' ]]; then
    brewtoken
  fi
  /usr/local/bin/brew $@
}
```

... and you first have to install the [JFrog CLI](https://jfrog.com/fr/getcli/) and execute `jfrog config add` to set your credentials up
(if you don't want to, you can rewrite the script above with simple curl calls)

### PR DESCRIPTION
**Pull request by zachauten**
...KEN env variable

- [x ] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x ] Have you successfully run `brew style` with your changes locally?
- [x ] Have you successfully run `brew typecheck` with your changes locally?
- [ x] Have you successfully run `brew tests` with your changes locally?

-----
I didn't see a unit test file for download_strategy.rb. If anyone knows where it's covered, I'll add some tests. I'd also be glad to add unit tests if coverage for that is missing altogether.

Closes #12574

### CODE REVIEW

File: Library/Homebrew/download_strategy.rb [hunk 1]
```diff
@@ -595,8 +595,16 @@ class CurlGitHubPackagesDownloadStrategy < CurlDownloadStrategy
   def initialize(url, name, version, **meta)
     meta ||= {}
     meta[:headers] ||= []
-    token = Homebrew::EnvConfig.artifact_domain ? Homebrew::EnvConfig.docker_registry_token : "QQ=="
-    meta[:headers] << "Authorization: Bearer #{token}" if token.present?
+    # token = Homebrew::EnvConfig.artifact_domain ? Homebrew::EnvConfig.docker_registry_token : "QQ=="
+    # meta[:headers] << "Authorization: Bearer #{token}" if token.present?
```
Comment (Rylan12) [inline Library/Homebrew/download_strategy.rb:599]:
```suggestion
```

File: Library/Homebrew/download_strategy.rb [hunk 2]
```diff
@@ -595,8 +595,16 @@ class CurlGitHubPackagesDownloadStrategy < CurlDownloadStrategy
...
+
+    meta[:headers] << if Homebrew::EnvConfig.artifact_domain && Homebrew::EnvConfig.docker_registry_token
+      "Authorization: Bearer #{Homebrew::EnvConfig.docker_registry_token}"
+    elsif Homebrew::EnvConfig.artifact_domain && Homebrew::EnvConfig.docker_registry_token_basic
+      "Authorization: Basic #{Homebrew::EnvConfig.docker_registry_token}"
```
Comment (Rylan12) [inline Library/Homebrew/download_strategy.rb:604]:
```suggestion
      "Authorization: Basic #{Homebrew::EnvConfig.docker_registry_token_basic}"
```

No?

**Review by Rylan12 — COMMENTED** [review]
There's some limited tests for various download strategy classes [here](https://github.com/Homebrew/brew/tree/master/Library/Homebrew/test/download_strategies). Don't see one for `CurlGitHubPackagesDownloadStrategy`, though.

File: Library/Homebrew/download_strategy.rb (hunk 2 above)
Comment (zachauten) [reply Library/Homebrew/download_strategy.rb:604]: 🤦‍♂️ yes, thanks!

File: Library/Homebrew/download_strategy.rb [hunk 3]
```diff
@@ -595,8 +595,16 @@ class CurlGitHubPackagesDownloadStrategy < CurlDownloadStrategy
...
+    else
+      "Authorization: Bearer QQ=="
```
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply Library/Homebrew/download_strategy.rb:606]: It's required by GitHub Packages.

File: Library/Homebrew/env_config.rb [hunk 4]
```diff
@@ -175,7 +175,12 @@ module EnvConfig
                      "(where bottles may be stored).",
       },
       HOMEBREW_DOCKER_REGISTRY_TOKEN:          {
-        description: "Use this bearer token for authenticating with a Docker registry proxying GitHub Packages.",
+        description: "Use this bearer token for authenticating with a Docker registry proxying GitHub Packages. " \
+                     "Preferred over HOMEBREW_DOCKER_REGISTRY_TOKEN_BASIC.",
+      },
+      HOMEBREW_DOCKER_REGISTRY_TOKEN_BASIC:    {
```
Comment (MikeMcQuaid) [inline Library/Homebrew/env_config.rb:181]:
```suggestion
      HOMEBREW_DOCKER_REGISTRY_BASIC_AUTH_TOKEN:    {
```

**Comment by zachauten** [comment]
Still working on those unit tests. Is there any other documentation that needs to be updated for adding a new env variable, or does everything from brew.sh, elsewhere get generated from env_config.rb?

File: Library/Homebrew/download_strategy.rb [hunk 5]
```diff
@@ -595,8 +595,13 @@ class CurlGitHubPackagesDownloadStrategy < CurlDownloadStrategy
...
     meta ||= {}
     meta[:headers] ||= []
-    token = Homebrew::EnvConfig.artifact_domain ? Homebrew::EnvConfig.docker_registry_token : "QQ=="
-    meta[:headers] << "Authorization: Bearer #{token}" if token.present?
+    meta[:headers] << if Homebrew::EnvConfig.artifact_domain && Homebrew::EnvConfig.docker_registry_token
+      "Authorization: Bearer #{Homebrew::EnvConfig.docker_registry_token}"
+    elsif Homebrew::EnvConfig.artifact_domain && Homebrew::EnvConfig.docker_registry_basic_auth_token
+      "Authorization: Basic #{Homebrew::EnvConfig.docker_registry_basic_auth_token}"
+    else
+      "Authorization: Bearer QQ=="
```
Comment (MikeMcQuaid) [inline Library/Homebrew/download_strategy.rb:603]:
```suggestion
      # This QQ== is needed for the no-auth GitHub Packages default.
      "Authorization: Bearer QQ=="
```

File: Library/Homebrew/env_config.rb [hunk 6]
```diff
@@ -55,161 +55,166 @@ module EnvConfig
...
                      "(where bottles may be stored).",
       },
-      HOMEBREW_DOCKER_REGISTRY_TOKEN:          {
-        description: "Use this bearer token for authenticating with a Docker registry proxying GitHub Packages.",
+      HOMEBREW_DOCKER_REGISTRY_TOKEN:            {
+        description: "Use this bearer token for authenticating with a Docker registry proxying GitHub Packages. " \
+                     "Preferred over HOMEBREW_DOCKER_REGISTRY_TOKEN_BASIC.",
       },
-      HOMEBREW_GITHUB_PACKAGES_USER:           {
+      HOMEBREW_DOCKER_REGISTRY_BASIC_AUTH_TOKEN: {
```
Comment (MikeMcQuaid) [inline Library/Homebrew/env_config.rb:181]: Can you move this before `HOMEBREW_DOCKER_REGISTRY_TOKEN` to they are sorted alphabetically?

**Review by MikeMcQuaid — APPROVED** [review]
Looks good, a few suggested tweaks. Don't think it needs a unit test, personally.

**Comment by zachauten** [comment]
@MikeMcQuaid Tweaked! I'll skip the tests then, unless anyone disagrees.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks again @zachauten!

**Pull request merged**